### ISSUE TITLE
Batch Renormalization via Keras Layer not working with TF 2+ and tf.function

### ISSUE THREAD

**Issue by georgwiese** [tensorflow/tensorflow#43732]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS 10.15.6
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): `v2.3.0-54-gfcc4b966f1 2.3.1`
- Python version: `3.6.2`
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**

Passing a Tensor derived from a `tf.keras.Input` to the `renorm_clipping` parameter of `tf.keras.layers.BatchNormalization` fails with an error like this:

```
_SymbolicException: Inputs to eager execution function cannot be Keras symbolic tensors, but found [<tf.Tensor 'RealDiv_7:0' shape=(None, 1) dtype=float32>, <tf.Tensor 'AddV2_2:0' shape=(None, 1) dtype=float32>, <tf.Tensor 'RealDiv_8:0' shape=(None, 1) dtype=float32>]
```

**Describe the expected behavior**

Since `renorm_clipping` expects Tensors, I would expect any Tensor to work, including those derived from `tf.keras.Input`.

**Standalone code to reproduce the issue**

[This notebook](https://gist.github.com/georgwiese/12d341bd131e2b16c8d4bb1e275289a6) shows a small example with TF 2.3.1. It creates a model with a step as input, so that batch renorm parameters can be computed from it. Please let me know if this is not the way the step should be passed to the model.

The model is then used in a `tf.function` (originally as part of a custom training loop, but simply running the forward pass reproduces the issue).

**Other info / logs**
See attached notebook for traceback.

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.3, nightly version (`2.4.0-dev20201004`) and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/34b820e2c0bd51504ef91dfca447f3f4/untitled421.ipynb). Thanks!

**Comment by nikitamaia** [comment]
Hi @georgwiese, have you looked at the error message that occurs in nightly? It provides more information as to what is going on. Specifically `in compilation of <function predict at 0x7f9442407ae8>, found return value of type <class 'tensorflow.python.keras.engine.keras_tensor.KerasTensor'>, which is not a Tensor.` You can see [in the docs here](https://www.tensorflow.org/api_docs/python/tf/keras/Input#returns) that 'Note that even if eager execution is enabled, Input produces a symbolic tensor (i.e. a placeholder)' 

As a workaround, using `model.predict_step(x)` in your tf.function instead seems to work. Have you tried that? You might also want to look into writing a custom layer.

**Comment by georgwiese** [comment]
@nikitamaia Thanks for your quick response!
`model.predict_step(x)` works because it runs the model in inference mode (in my original code, if I set `training=False`, I also get no error; same if the `BatchNormalization` parameters do not depend on `step`).

For the actual use case, I want to train the model and hence need to run the model in training mode. It is trained using a custom training loop, but I could also reproduce the same error using `tf.keras.Model.fit()` and `tf.keras.Model.train_step()`. Please see [this Notebook](https://gist.github.com/georgwiese/eaa9f048e2a85d93f0bec89c91949cef).

So, is there any way to currently use batch renormalization with a dynamic schedule without re-implementing it as a custom layer?

**Comment by nikitamaia** [comment]
Ah yes, makes sense why I didn't see the error message with `model.predict_step(x)`!
However, I think this question is a better fit for Stack Overflow since there's a bigger community to help. Github is best for bugs/performance issues.

**Comment by georgwiese** [comment]
I [have posted this on Stack Overflow](https://stackoverflow.com/questions/64155323/batch-renormalization-with-tensorflow-2-and-tf-function), but I'm becoming convinced that this is a bug, or at least a documentation issue. It looks like there is no way to use batch renormalization in the recommended way (i.e., with a dynamic schedule).

Or am I doing something unconventional in the notebook linked above?

**Comment by gowthamkpr** [comment]
@georgwiese Tried running your code with latest version of tensorflow (tf 2.8.0). Ran into a different error. Here's the [colab](https://colab.research.google.com/gist/gowthamkpr/3c0a1349de501cf89d15a62f06c0dae8/untitled.ipynb). Pls take a look at the solution provided [here](https://stackoverflow.com/questions/61910143/how-to-use-tf-function-with-keras-sequential-apis) to solve the new error.
